### ISSUE TITLE
QueueService class in azure-storage 0.34.3 not pickleable, so unusuable in a Python MP context

### ISSUE THREAD

**Issue by lstavenhagen** [Azure/azure-storage-python#317]
Hi,
I originally posted this in the azure-sdk-for-python repository, and was told this was the proper place.

Instances of the QueueService class in the azure-storage 0.34.3 repository appear to not be serializable by the pickle module in Python 2.7, and therefore it can't be used in a Python MP context (eg. in Process objects). The default decoding/encoding callback functions appear to be implemented in a static class QueueMessageFormat using @staticmethod methods, which appear not to be pickleable. 
A workaround is to implement custom "non-static" callbacks and set the instance fields in the QueueService object to these custom callbacks.

Perhaps this design pattern should be changed so the QueueService class is pickleable by default, or guidance could be given to implement your own encode/decode callbacks (forgive me if this is already documented).

Platform: Windows
Python version: 2.7.9 +

Reproducible source code scenario distilled from SDK code:

```python
import pickle

class QueueMessageFormat:
    @staticmethod
    def noencode(data):
        return data     
    @staticmethod
    def nodecode(data):
        return data
class QueueService(object):
    def __init__(self):
        self.encode_function = QueueMessageFormat.noencode
        self.decode_function = QueueMessageFormat.nodecode
def main():
    f=QueueService()
    with open("pickleme.txt","wb") as output:
        pickle.dump(f,output)
if __name__ == "__main__":
    main()
```
This example yields "pickle.PicklingError: Can't pickle <function noencode at 0x02880930>: it's not found as __main__.noencode"

Azure SDK install information ("pip freeze azure")
adal==0.4.5
asn1crypto==0.22.0
azure==2.0.0
azure-batch==3.0.0
azure-common==1.1.6
azure-datalake-store==0.0.12
azure-graphrbac==0.30.0
azure-keyvault==0.3.5
azure-mgmt==1.0.0
azure-mgmt-authorization==0.30.0
azure-mgmt-batch==4.0.0
azure-mgmt-cdn==0.30.3
azure-mgmt-cognitiveservices==1.0.0
azure-mgmt-compute==1.0.0
azure-mgmt-containerregistry==0.2.1
azure-mgmt-datalake-analytics==0.1.6
azure-mgmt-datalake-nspkg==2.0.0
azure-mgmt-datalake-store==0.1.6
azure-mgmt-devtestlabs==2.0.0
azure-mgmt-dns==1.0.1
azure-mgmt-documentdb==0.1.3
azure-mgmt-iothub==0.2.2
azure-mgmt-keyvault==0.31.0
azure-mgmt-logic==2.1.0
azure-mgmt-monitor==0.2.1
azure-mgmt-network==1.0.0
azure-mgmt-nspkg==2.0.0
azure-mgmt-rdbms==0.1.0
azure-mgmt-redis==4.1.0
azure-mgmt-resource==1.1.0
azure-mgmt-scheduler==1.1.2
azure-mgmt-sql==0.5.3
azure-mgmt-storage==1.0.0
azure-mgmt-trafficmanager==0.30.0
azure-mgmt-web==0.32.0
azure-nspkg==2.0.0
azure-servicebus==0.21.1
azure-servicefabric==5.6.130
azure-servicemanagement-legacy==0.20.6
azure-storage==0.34.3
certifi==2017.4.17
cffi==1.10.0
chardet==3.0.4
cryptography==1.9
enum34==1.1.6
futures==3.1.1
idna==2.5
ipaddress==1.0.18
isodate==0.5.4
keyring==10.4.0
msrest==0.4.11
msrestazure==0.4.10
oauthlib==2.0.2
pathlib2==2.3.0
pycparser==2.17
PyJWT==1.5.2
python-dateutil==2.6.0
pywin32-ctypes==0.1.2
requests==2.18.1
requests-oauthlib==0.8.0
scandir==1.5
six==1.10.0
urllib3==1.21.1

**Comment by zezha-msft** [comment]
Hi @lstavenhagen, I apologize for the delayed response. 😭 

If I'm understanding this correctly, you are trying to exchange a QueueService object between processes, right? Since the QueueService class does not contain much state, I think a very easy workaround would be writing your own class to contain the credentials and other info that would normally be passed to the constructor of QueueService, and exchange an instance of this class between your processes. Then each process would create their own instance of QueueService.

Please let me know if this works. If not, I'll reopen this. 😄

**Comment by lstavenhagen** [comment]
Hi Zezha,
Yes, that's exactly what my code does, in fact. I have a wrapper class that encapsulates instantiating a QueueService class object, with thin wrappers around the methods of the class to abstract any interface changes, slightly customize behavior, etc. 

An object of that wrapper class is used with multiple feeders created as a pool of Python Process objects. 

But that's exactly what no longer works in the current azure-storage module that used to work in 0.20.0 (IIRC), and the change in the code responsible is in the implementation of the encoder/decoder callbacks as I schematized above. 

To me, that's buggy, since multiple feeders on a QueueService instance is kind of the point of it, even on multiprocess VM's out in Azure.... 

Thank you,
LS

**Comment by zezha-msft** [comment]
Hi @lstavenhagen,

I am a bit confused. If QueueService is instantiated on each process by your wrapper class, then why would there by any problem? In other words, if you never pickle QueueService but only the wrapper class containing the credentials, then you shouldn't have any problem, right?

I agree that it would be nice to be able to pickle QueueService directly, but unfortunately we did not consider this scenario. I think that pickling your wrapper object is definitely the easier way. In addition, since you are already creating multiple processes, which are expensive, creating a new QueueService on each process should be negligible in comparison.

**Comment by lstavenhagen** [comment]
No, there is only one instance of the QueueService class in my scenario. That instance, however, is fed upon by multiple Python processes. Basically I do: 

q_object=MyWrapperClassWhichEncapsulatesQueueService(...)

enqueue_utility(q_object,...)

for i in range(cpus_on_machine):
    processes.append(Process(target=mp_queue_feeder_function, args=(q_object,...)))
for p in processes:
    p.start()
for p in processes:
    p.join()

So it's a single instance that's used by multiple processes in Python on a single machine. Of course, this works too with multiple machines running only one process/thread feeding on a QueueService object, but we want to take advantage of multiple processes within a single VM as well. 

And we just got used to it working in 0.20.0 also so :).

Thank you,
LS

**Comment by zezha-msft** [comment]
@lstavenhagen  Ah OK I see.

So what is stopping you from creating a new instance of QueueService on each process? That solves the pickling problem right? QueueService contains only the credentials and configurations, it is stateless. So I do not see why you have to share the same instance of QueueService on all the processes, besides saving on memory. Please let me know. 😄  thanks!

**Comment by lstavenhagen** [comment]
Well, the queue is populated with all the "tasks" to be run by the main program (represented by "enqueue_utility(q_object,...) beforehand; the individual processes consume the single queue in a multi-process fashion.
If I instantiate a seperate queue for each process, that would involve parallizing the enqueing itself, somehow, which was a lot more complicated than using just a single queue object. So I just instantiated one queue that's populated with the entire task set in the main process before the MP consumption was done, that all the processes feed on.

Thank you,
LS

**Comment by zezha-msft** [comment]
Hi @lstavenhagen, I understand that you are maintaining a list of tasks using your q_object, and q_object wraps QueueService to make service calls to Azure Storage. What I am trying to convey here is that you do not have to share the inner QueueService object on q_object with all the processes, doing so causes your problem, since pickling does not work on QueueService.

I have no knowledge of what your wrapper class is doing exactly, but I believe doing something like this, by avoiding creating the QueueService object on q_object, should solve your problem:

``` Python
class WrapperClass:
    account_key = 'bla'
    account_name = 'bla'

    def _initialize_queue_service(self):
            return QueueService(account_name, account_key)

    def get_messages(self):
        queue_service = self._initialize_queue_service()
        # do other stuffs
        # call queue_service to access service
```

I admit that this is not a particularly elegant workaround, as it requires you to create a new instance of QueueService in every method that make a service call. However, given your situation, I believe it is tolerable. Alternatively, you could modify our library to make the QueueService pickable, as this is after all an open source project. You could even submit a PR for it! 😄 

Please let me know if you have any other concern, thanks!

**Comment by lstavenhagen** [comment]
Hi Zezha,
Ok I think I see what you mean - Interesting proposal. But are you saying there's a way to, say, connect to an existing queue object initialized in a particular storage account by some separate application/script and start feeding from it?  I looked through the SDK source in queueservice.py, and nothing jumped out at me that looked like support for that. 

But I think that's what you're proposing, something like:

Main_setup_process.py:
main_q_service=QueueService(...)
global_q_handle=main_q_service.create_queue("globalqueue",...)
populate(global_q_handle)
    for test in testset:
        global_q_handle.put_message("globalqueue",...)
...

And then somehow in a subprocess, something like what you proposed:

worker_role.py:
local_q_service=QueueService(...)
local_q=local_q_service.connect_to_queue_and_return_handle("globalqueue",)
start_feeding(local_q)
...
If it can be done this way, without a global Python QueueService() instance with state in it etc., but instead a way to create the queue, populate it and that's it in a setup process, then a way to connect via REST/SDK calls to the queue object itself out in the Azure storage account where it was initialized, that would be a different design pattern that I think would work.

But it's not obvious to me that the SDK has any support for that - and I don't know the queue services REST API well enough to say if that's how it intends queue services to be used.

Is that possible? If so, that's an intriguing idea....

Thank you,
LS

**Comment by lstavenhagen** [comment]
PS: ah, unless this is already how get_message(), put_message(), etc. already work..... they just go by the queue name and reference the account info passed into the QueueService() constructor?

If thats the case, I should probably change my design rather than muck up ya'lls SDK, which is already super nice :). I ought to be able to get rid of the global Python instance pretty easily...

Let me know if I'm interpreting that right...
LS

**Comment by lstavenhagen** [comment]
Ok, I've deleted my previous comments. You're right and I can redesign to accommodate this in the manner you describe here. I just have to refactor my wrapper class into one that initializes a queue and another that uses the queue by name via a seperate QueueService() instance in each subprocess. 

I wrote this code 2 years ago when I was still learning stateless programming in Azure :). So I think my design should be redone instead of hacking the SDK :).

Thank you,
LS
